File: sms-service/src/main/java/com/iiplabs/nns/sms/services/SmsService.java
package com.iiplabs.nns.sms.services;

import java.util.Random;
import java.util.concurrent.TimeUnit;

import com.iiplabs.nns.sms.exceptions.SmsServiceException;
import com.iiplabs.nns.sms.model.dto.SendMessageRequestDto;

import org.springframework.stereotype.Service;

import lombok.extern.log4j.Log4j2;

@Log4j2
@Service
public class SmsService implements ISmsService {

  private static final int MAX_EXECUTION_DELAY = 5;

  @Override
  public void send(SendMessageRequestDto sendMessageRequestDto) {
    log.info(String.format("Sending SMS to %s that %s is now available", sendMessageRequestDto.getDestinationPhone(),
        sendMessageRequestDto.getSourcePhone()));
    Random random = new Random(System.nanoTime());

    try {
      TimeUnit.SECONDS.sleep(random.nextInt(MAX_EXECUTION_DELAY));
    } catch (InterruptedException ie) {
      Thread.currentThread().interrupt();
      throw new SmsServiceException(ie.getMessage());
    }

    log.info(String.format("Success sending SMS to %s", sendMessageRequestDto.getDestinationPhone()));
  }

}

File: ksms-service/src/main/kotlin/com/iiplabs/nns/ksms/service/SmsService.kt
package com.iiplabs.nns.ksms.service

import mu.KotlinLogging
import kotlinx.coroutines.delay
import org.springframework.stereotype.Service

import com.iiplabs.nns.ksms.model.SendMessageRequestDto
import kotlin.time.Duration.Companion.seconds

interface SmsService {
    suspend fun send(sendMessageRequestDto: SendMessageRequestDto)
}

private val logger = KotlinLogging.logger {}

@Service("smsService")
class SmsServiceImpl : SmsService {

    companion object {
        private val MAX_EXECUTION_DELAY = 5L
    }

    override suspend fun send(sendMessageRequestDto: SendMessageRequestDto) {
        logger.info { "Sending SMS to ${sendMessageRequestDto.destinationPhone} that ${sendMessageRequestDto.sourcePhone} is now available" }

        val randomDelayLong = (1..MAX_EXECUTION_DELAY).shuffled().first().toLong()
        delay(randomDelayLong.seconds)

        logger.info { "Success sending SMS to ${sendMessageRequestDto.destinationPhone}, with delay ${randomDelayLong} seconds" }
    }

}
